File: backend/cache/smart_cache.py
"""
Smart Cache Implementation
Flagship Adaptive, Application-Aware Cache with Multi-Factor Scoring, Dynamic TTL, and Pollution Defense.
"""

from typing import Dict, Any, Optional, List
from backend.core.types import CacheStrategy, DecisionType, WorkloadType, SystemConfig
from backend.cache.base_cache import BaseCache, CacheEntry
from backend.cache.smart_scorer import SmartScorer
from backend.cache.ttl_manager import DynamicTTLManager
from backend.cache.pollution_guard import PollutionGuard


class SmartCache(BaseCache):
    def __init__(
        self,
        capacity_bytes: int = SystemConfig.cacheCapacityBytes,
        workload_type: WorkloadType = WorkloadType.READ_HEAVY_API
    ):
        super().__init__(CacheStrategy.SMART, capacity_bytes)
        self.scorer = SmartScorer(workload_type)
        self.ttl_manager = DynamicTTLManager()
        self.pollution_guard = PollutionGuard()
        self.max_item_size_bytes = SystemConfig.maxItemSizeBytes
        self.window_requests = 0
        self.workload_type = workload_type

    def set_workload_type(self, workload_type: WorkloadType):
        self.workload_type = workload_type
        self.scorer.update_weights_for_workload(workload_type)

    def get(self, id: str, current_time: float) -> Dict[str, Any]:
        entry = self.entries.get(id)
        if not entry:
            self.misses += 1
            return {"hit": False, "entry": None}

        age = current_time - entry.created_at

        # Check for expiration
        if age >= entry.current_ttl:
            # Expired
            del self.entries[id]
            self.used_bytes -= entry.size_bytes
            self.misses += 1
            return {"hit": False, "entry": None, "expired": True}

        # HIT
        self.hits += 1
        entry.total_hits += 1
        entry.recent_hits += 1
        entry.last_accessed_at = current_time

        # Check proactive REFRESH
        if age >= entry.current_ttl * 0.75 and entry.score >= 0.45:
            entry.decision = DecisionType.REFRESH.value
            entry.reason = (
                f"Approaching TTL expiration ({round(age)}s / {entry.current_ttl}s). "
                f"Proactive background refresh triggered (Score {round(entry.score, 2)}) to prevent miss stampede."
            )
            self.refreshes += 1
            entry.created_at = current_time  # Refreshed timestamp

        return {"hit": True, "entry": entry}

    def put(self, item: Dict[str, Any], current_time: float) -> Dict[str, Any]:
        item_id = item["id"]
        size_bytes = item.get("sizeBytes", 4096)
        is_pollution = item.get("isPollutionKey", False)

        # Reject single item exceeding full cache
        if size_bytes > self.capacity_bytes:
            return {"admitted": False, "reason": "Object size exceeds total cache capacity"}

        # Pollution Guard evaluation
        self.pollution_guard.record_request(item_id, is_pollution, current_time)
        pollution_status = self.pollution_guard.evaluate(self.get_entries_list(), current_time)
        
        if pollution_status["riskLevel"] == "HIGH" and is_pollution:
            return {"admitted": False, "reason": "Rejected by Cache Pollution Defense: low-reuse crawler key."}

        # Update if already exists
        if item_id in self.entries:
            existing = self.entries[item_id]
            self.used_bytes -= existing.size_bytes
            existing.last_accessed_at = current_time
            existing.total_hits += 1
            existing.recent_hits += 1
            existing.size_bytes = size_bytes
            self.used_bytes += size_bytes
            self.recalculate_entry(existing, current_time)
            return {"admitted": True, "entry": existing}

        # Evict space if needed
        bytes_needed = (self.used_bytes + size_bytes) - self.capacity_bytes
        if bytes_needed > 0:
            self.evict(bytes_needed, current_time)

        # Create new entry
        entry = CacheEntry(
            id=item_id,
            name=item.get("name", item_id),
            category=item.get("category", "General"),
            item_type=item.get("type", "DATA"),
            size_bytes=size_bytes,
            base_db_latency_ms=item.get("baseDbLatencyMs", 50.0),
            recompute_cost_units=item.get("recomputeCostUnits", 1.0),
            update_volatility=item.get("updateVolatility", 0.1),
            created_at=current_time,
            current_ttl=SystemConfig.defaultTTL
        )

        self.entries[item_id] = entry
        self.used_bytes += size_bytes
        self.recalculate_entry(entry, current_time)

        return {"admitted": True, "entry": entry}

    def recalculate_entry(self, entry: CacheEntry, current_time: float) -> CacheEntry:
        context = {
            "current_time": current_time,
            "cache_capacity_bytes": self.capacity_bytes,
            "max_item_size_bytes": self.max_item_size_bytes,
            "total_window_requests": max(1, self.window_requests)
        }

        eval_res = self.scorer.evaluate_object(entry, context)
        ttl_res = self.ttl_manager.compute_ttl(entry, eval_res)

        entry.score = eval_res["finalScore"]
        entry.factors = eval_res["factors"]
        entry.weights = eval_res["weights"]
        entry.prev_ttl = ttl_res["prevTTL"]
        entry.current_ttl = ttl_res["newTTL"]

        age = current_time - entry.created_at

        # Derive explainable decision
        if age >= entry.current_ttl * 0.80 and entry.score >= 0.45:
            entry.decision = DecisionType.REFRESH.value
            entry.reason = (
                f"High-demand item ({entry.total_hits} hits, score {entry.score:.2f}) "
                f"approaching staleness threshold ({int(age)}s / {entry.current_ttl}s). Proactive refresh active."
            )
        elif entry.score >= 0.40:
            entry.decision = DecisionType.RETAIN.value
            entry.reason = (
                f"High frequency ({entry.recent_hits} hits) + retrieval cost ({entry.base_db_latency_ms}ms) "
                f"outweighs footprint. Value Score: {entry.score:.2f}."
            )
        else:
            entry.decision = DecisionType.EVICT.value
            entry.reason = (
                f"Low utility/reuse probability (Score {entry.score:.2f}) "
                f"prioritizes this object for eviction under memory pressure."
            )

        return entry

    def evict(self, bytes_needed: int, current_time: float) -> int:
        freed_bytes = 0
        entries_list = self.get_entries_list()

        # Recalculate all resident entries
        for e in entries_list:
            self.recalculate_entry(e, current_time)

        # Sort: protected items placed at back, lowest score first
        def sort_key(e: CacheEntry):
            is_prot = 1 if self.pollution_guard.is_protected(e.id) else 0
            return (is_prot, e.score)

        entries_list.sort(key=sort_key)

        for victim in entries_list:
            if freed_bytes >= bytes_needed:
                break
            if victim.id in self.entries:
                del self.entries[victim.id]
                self.used_bytes -= victim.size_bytes
                freed_bytes += victim.size_bytes
                self.evictions += 1

        return freed_bytes

    def periodic_maintenance(self, current_time: float, total_window_requests: int):
        self.window_requests = total_window_requests
        self.pollution_guard.evaluate(self.get_entries_list(), current_time)

        for entry in self.entries.values():
            entry.prev_window_hits = entry.recent_hits
            entry.recent_hits = int(entry.recent_hits * 0.7)  # Exponential decay
            self.recalculate_entry(entry, current_time)

File: src/cache/smart-cache.js
/**
 * Smart Cache Implementation
 * Flagship Adaptive, Application-Aware Cache with Multi-Factor Scoring, Dynamic TTL, and Pollution Defense
 */

import { BaseCache } from './base-cache.js';
import { SmartScorer } from './smart-scorer.js';
import { DynamicTTLManager } from './ttl-manager.js';
import { PollutionGuard } from './pollution-guard.js';
import { CACHE_STRATEGIES, DECISION_TYPES, DEFAULT_CONFIG } from '../core/types.js';

export class SmartCache extends BaseCache {
  constructor(capacityBytes = DEFAULT_CONFIG.cacheCapacityBytes, workloadType) {
    super(CACHE_STRATEGIES.SMART, capacityBytes);
    this.scorer = new SmartScorer(workloadType);
    this.ttlManager = new DynamicTTLManager();
    this.pollutionGuard = new PollutionGuard();
    this.maxItemSizeBytes = DEFAULT_CONFIG.maxItemSizeBytes;
    
    // Activity tracking for rate calculation
    this.windowRequests = 0;
    this.lastEvaluatedAt = 0;
  }

  setWorkloadType(workloadType) {
    this.scorer.updateWeightsForWorkload(workloadType);
  }

  get(id, currentTime) {
    const entry = this.entries.get(id);
    if (!entry) {
      this.misses++;
      return { hit: false, entry: null };
    }

    const age = currentTime - entry.createdAt;
    
    // Check for expiration
    if (age >= entry.currentTTL) {
      // Expired
      this.entries.delete(id);
      this.usedBytes -= entry.sizeBytes;
      this.misses++;
      return { hit: false, entry: null, expired: true };
    }

    // HIT
    this.hits++;
    entry.totalHits++;
    entry.recentHits = (entry.recentHits || 0) + 1;
    entry.lastAccessedAt = currentTime;

    // Check if item qualifies for proactive REFRESH (approaching staleness but high utility)
    if (age >= entry.currentTTL * 0.75 && entry.score >= 0.45) {
      entry.decision = DECISION_TYPES.REFRESH;
      entry.reason = `Approaching TTL threshold (${Math.round(age)}s / ${entry.currentTTL}s). Background refresh triggered to preserve hot cache line.`;
      this.refreshes++;
      entry.createdAt = currentTime; // Proactively refreshed
    }

    return { hit: true, entry };
  }

  put(item, currentTime) {
    const sizeBytes = item.sizeBytes || 4096;

    // Reject single item exceeding full cache
    if (sizeBytes > this.capacityBytes) {
      return { admitted: false, reason: 'Object size exceeds total cache capacity' };
    }

    // Record with Pollution Guard
    this.pollutionGuard.recordRequest(item.id, item.isPollutionKey, currentTime);

    // If pollution attack is high and item is an unverified single-use key, reject or quarantine
    const pollutionStatus = this.pollutionGuard.evaluate(this.getEntriesList(), currentTime);
    if (pollutionStatus.riskLevel === 'HIGH' && item.isPollutionKey) {
      return { admitted: false, reason: 'Rejected by Cache Pollution Defense: low-reuse ephemeral key.' };
    }

    // If already exists, update entry
    if (this.entries.has(item.id)) {
      const existing = this.entries.get(item.id);
      this.usedBytes -= existing.sizeBytes;
      existing.lastAccessedAt = currentTime;
      existing.totalHits++;
      existing.recentHits = (existing.recentHits || 0) + 1;
      existing.sizeBytes = sizeBytes;
      this.usedBytes += sizeBytes;
      this.recalculateEntry(existing, currentTime);
      return { admitted: true, entry: existing };
    }

    // Evict space if needed
    const bytesNeeded = (this.usedBytes + sizeBytes) - this.capacityBytes;
    if (bytesNeeded > 0) {
      this.evict(bytesNeeded, currentTime);
    }

    // Create fresh entry
    const entry = {
      id: item.id,
      name: item.name,
      category: item.category,
      type: item.type,
      sizeBytes,
      baseDbLatencyMs: item.baseDbLatencyMs || 50,
      recomputeCostUnits: item.recomputeCostUnits || 1.0,
      updateVolatility: item.updateVolatility || 0.1,
      createdAt: currentTime,
      lastAccessedAt: currentTime,
      totalHits: 1,
      recentHits: 1,
      prevWindowHits: 0,
      currentTTL: DEFAULT_CONFIG.defaultTTL,
      prevTTL: DEFAULT_CONFIG.defaultTTL,
      score: 0.50,
      decision: DECISION_TYPES.RETAIN,
      reason: 'Newly admitted into cache store.',
      factors: {}
    };

    this.entries.set(item.id, entry);
    this.usedBytes += sizeBytes;

    // Immediate scoring evaluation
    this.recalculateEntry(entry, currentTime);

    return { admitted: true, entry };
  }

  recalculateEntry(entry, currentTime) {
    const context = {
      currentTime,
      cacheCapacityBytes: this.capacityBytes,
      maxItemSizeBytes: this.maxItemSizeBytes,
      totalWindowRequests: Math.max(1, this.windowRequests)
    };

    const evalResult = this.scorer.evaluateObject(entry, context);
    const ttlResult = this.ttlManager.computeTTL(entry, evalResult);

    entry.score = evalResult.finalScore;
    entry.factors = evalResult.factors;
    entry.weights = evalResult.weights;
    entry.prevTTL = ttlResult.prevTTL;
    entry.currentTTL = ttlResult.newTTL;

    const age = currentTime - entry.createdAt;

    // Compute explicit decision and explainable reason
    if (age >= entry.currentTTL * 0.80 && entry.score >= 0.45) {
      entry.decision = DECISION_TYPES.REFRESH;
      entry.reason = `High demand item (${entry.totalHits} hits, score ${entry.score}) approaching staleness threshold. Proactive refresh active.`;
    } else if (entry.score >= 0.40) {
      entry.decision = DECISION_TYPES.RETAIN;
      entry.reason = `High frequency + retrieval cost (${entry.baseDbLatencyMs}ms) + recent access justifies cache occupancy.`;
    } else {
      entry.decision = DECISION_TYPES.EVICT;
      entry.reason = `Low reuse probability + score (${entry.score}) makes this object prime candidate for eviction under memory pressure.`;
    }

    return entry;
  }

  evict(bytesNeeded, currentTime) {
    let freedBytes = 0;

    // Recalculate all entries before eviction ranking
    const entriesList = this.getEntriesList();
    for (const e of entriesList) {
      this.recalculateEntry(e, currentTime);
    }

    // Sort by Score ascending (lowest score evicted first), protecting shielded items
    entriesList.sort((a, b) => {
      const aProtected = this.pollutionGuard.isProtected(a.id) ? 1 : 0;
      const bProtected = this.pollutionGuard.isProtected(b.id) ? 1 : 0;
      if (aProtected !== bProtected) return aProtected - bProtected;
      return a.score - b.score;
    });

    for (const victim of entriesList) {
      if (freedBytes >= bytesNeeded) break;
      this.entries.delete(victim.id);
      this.usedBytes -= victim.sizeBytes;
      freedBytes += victim.sizeBytes;
      this.evictions++;
    }

    return freedBytes;
  }

  periodicMaintenance(currentTime, totalWindowRequests) {
    this.windowRequests = totalWindowRequests;
    this.pollutionGuard.evaluate(this.getEntriesList(), currentTime);

    // Update hits and decay
    for (const entry of this.entries.values()) {
      entry.prevWindowHits = entry.recentHits || 0;
      entry.recentHits = Math.floor((entry.recentHits || 0) * 0.7); // decay
      this.recalculateEntry(entry, currentTime);
    }
  }
}
